[InstCombine] Improve coverage of `foldSelectValueEquivalence`

goldsteinn opened the pull request:
- **[InstCombine] Add tests for expanding `foldSelectValueEquivalence`; NFC**
- **[InstCombine] Improve coverage of `foldSelectValueEquivalence`**

1) We don't need the `noundef` check if the new simplification is a
constant and doesn't contain undefs.

This cleans up regressions from folding multiuse:
    `(icmp eq/ne (sub/xor x, y), 0)` -> `(icmp eq/ne x, y)`.
in #87275

goldsteinn committed:
[InstCombine] Add tests for expanding `foldSelectValueEquivalence`; NFC

goldsteinn commented:
ping

goldsteinn commented:
ping

dtcxzyw commented:
Can you copy the commit message into the PR description? I know you always merge your patchs locally. But it may be more convenient for me to review :)

dtcxzyw commented on llvm/lib/Transforms/InstCombine/InstCombineSelect.cpp:
```diff
@@ -1285,40 +1285,97 @@ Instruction *InstCombinerImpl::foldSelectValueEquivalence(SelectInst &Sel,
+    };
 
+    if (Value *V = simplifyWithOpReplaced(TrueVal, OldOp, NewOp, SQ,
+                                          /* AllowRefinement */ true)) {
```
```suggestion
                                          /*AllowRefinement=*/ true)) {
```

dtcxzyw commented on llvm/lib/Transforms/InstCombine/InstCombineSelect.cpp:
```diff
@@ -1285,40 +1285,97 @@ Instruction *InstCombinerImpl::foldSelectValueEquivalence(SelectInst &Sel,
 
+    if (Value *V = simplifyWithOpReplaced(TrueVal, OldOp, NewOp, SQ,
+                                          /* AllowRefinement */ true)) {
+      // Need some gurantees about the new simplified op to ensure we don't inf
```
```suggestion
      // Need some guarantees about the new simplified op to ensure we don't inf
```

dtcxzyw commented on llvm/lib/Transforms/InstCombine/InstCombineSelect.cpp:
```diff
@@ -1285,40 +1285,97 @@ Instruction *InstCombinerImpl::foldSelectValueEquivalence(SelectInst &Sel,
+    //     miss a one-use simplification.
+    //  2) Create a new one-use instruction.
+    // proceed.
+    else if (TrueVal->hasOneUse() &&
```
I think this `else` can be dropped because we can fall through to this line if the above simplification failed.

goldsteinn commented:
ping

dtcxzyw reviewed (commented):
This patch is too complicated to me. I can do some tests to see the ir diff.

dtcxzyw commented:
> This patch is too complicated to me. I can do some tests to see the ir diff.

Unfortunately, I have to upgrade my datasets to remove icmp/fcmp constant expressions :(

goldsteinn commented:
> This patch is too complicated to me. I can do some tests to see the ir diff.

Would it help if I split it into multiple commits? If not any suggestions for simplifying to make it reviewable?

dtcxzyw commented:
> If not any suggestions for simplifying to make it reviewable?

I guess I can provide some suggestions after the ir diff is available :)

goldsteinn commented:
> > If not any suggestions for simplifying to make it reviewable?
> 
> I guess I can provide some suggestions after the ir diff is available :)

Okay, thanks

nikic commented:
What was the motivating case from https://github.com/llvm/llvm-project/pull/87275?

goldsteinn commented:
> What was the motivating case from #87275?

Yes, although we are waiting to get @dtcxzyw's results to see if it has other benefits.

nikic commented:
Sorry, what I meant is, which of the test cases / changes in this PR is the one that originally motivated this patch?

goldsteinn commented:
> > > If not any suggestions for simplifying to make it reviewable?
> > 
> > 
> > I guess I can provide some suggestions after the ir diff is available :)
> 
> See https://alive2.llvm.org/ce/z/EUTKhF. I think we can do this simplification iff `simplifyWithOpReplaced` returns a constant.

Looking at the diff, it seems we get some benefit from the non-const simplification.
Ill split to multiple commits. If that doesn't simplify review ill split to multiple PRs.

goldsteinn committed:
[InstCombine] Improve coverage of `foldSelectValueEquivalence` for constants

We don't need the `noundef` check if the new simplification is a
constant.

This cleans up regressions from folding multiuse:
    `(icmp eq/ne (sub/xor x, y), 0)` -> `(icmp eq/ne x, y)`.

goldsteinn commented:
@nikic and @dtcxzyw

I simplified the code a bit. Hopefully that helps. Also I split changes for folding constants / non-constants into two seperate commits.

If its still too complicated I will split to 2 seperate PRs. That said I hope for both of the changes to get in. The non-const change seems to be the more common in @dtcxzyw's tests and seems to have independent value.

dtcxzyw reviewed (approved):
The first two commits LGTM. Please file a separate PR for the remaining changes. If there are some problems in the third commit, we don't need to revert all the commits.

dtcxzyw commented on llvm/lib/Transforms/InstCombine/InstCombineSelect.cpp:
```diff
@@ -1308,11 +1310,52 @@ Instruction *InstCombinerImpl::foldSelectValueEquivalence(SelectInst &Sel,
+
+      // If we simplified the TrueArm -> NewOp then replace.
+      // This handles things like `select`/`min`/`max`/`or`/`and`/etc...
+      if (NewOp == V)
```
Can you do some stress tests (e.g., csmith) for 3-4 days (~10 million test cases on a 128C server)?

goldsteinn replied:
Hmm, maybe a case like `(select (icmp eq (or x, y), y), (or x, y), Q)`. Where we replace `y` with `(or x, y)` and then `(or x, y), x/y)` -> `(or x, y)`.

I think for it to inf loop through would need `NewOp == TrueArm`. I think we can probably safely just return null for that case.

Either way, ill run csmith for a bit.

What do you mean 128C server?

v01dXYZ replied:
Maybe 128C == 128 Cores ?
Could be 128 C° though.

nikic commented on llvm/lib/Transforms/InstCombine/InstCombineSelect.cpp:
```diff
@@ -1288,40 +1288,94 @@ Instruction *InstCombinerImpl::foldSelectValueEquivalence(SelectInst &Sel,
+    bool NewOpNeverUndef = false;
+
+    if (Value *V = simplifyWithOpReplaced(TrueVal, OldOp, NewOp, SQ,
+                                          /* AllowRefinement= */ true)) {
```
```suggestion
                                          /* AllowRefinement=*/true)) {
```

nikic reviewed (commented):
I think the second commit is not quite correct. I'll try to come up with an example.

nikic commented:
See select_replacement_add_eq_vec_undef for the problematic case. The issue is with undef values, not poison values, and that problem exists even if the result is a constant. I've clarified the implementation to that effect.

goldsteinn commented:
> See select_replacement_add_eq_vec_undef for the problematic case. The issue is with undef values, not poison values, and that problem exists even if the result is a constant. I've clarified the implementation to that effect.

So I think this can be fixed if the constant result doesn't contain any undefs.

nikic reviewed (commented):
Can you please drop the last commit and its tests from this PR?

goldsteinn commented:
> Can you please drop the last commit and its tests from this PR?

Done

nikic commented on llvm/lib/Transforms/InstCombine/InstCombineSelect.cpp:
```diff
@@ -1288,20 +1288,36 @@ Instruction *InstCombinerImpl::foldSelectValueEquivalence(SelectInst &Sel,
+      // Need some guarantees about the new simplified op to ensure we don't inf
+      // loop.
+      // If we simplify to a constant, replace if we aren't creating new undef.
+      if (match(V, m_ImmConstant())) {
+        if (isGuaranteedNotToBeUndef(V, SQ.AC, &Sel, &DT) ||
+            isGuaranteedNotToBeUndef(NewOp, SQ.AC, &Sel, &DT))
+          return replaceOperand(Sel, Swapped ? 2 : 1, V);
+        return nullptr;
+      }
```
```suggestion
      if (match(V, m_ImmConstant()) &&
          isGuaranteedNotToBeUndef(V, SQ.AC, &Sel, &DT))
        return replaceOperand(Sel, Swapped ? 2 : 1, V);
```
Would all tests still pass if you did this? I think you added `select_replacement_add_eq_vec_undef_okay` to cover that second isGuaranteedNotToBeUndef call, but I think that test would also be covered by falling through to the following case. I'd also be fine with just regressing that one entirely, as the folding result being a constant with undef isn't something we expect to see in practice.

goldsteinn replied:
I think you're right. Let me try and update.

goldsteinn replied:
Updated.

Going to push this in a bit.

nikic reviewed (approved):
LGTM, thanks!
